src/service/localization/index.js

```diff
@@ -144,4 +152,57 @@ export class LocalizationService {
       localizedStringId
     );
   }
+
+  /**
+   * Resolves with the localized string when the registered bundles contain them.
+   * @param {!LocalizedStringId_Enum} localizedStringId
+   * @param {!Element=} elementToUse
+   * @return {!Promise{?string}}
+   * @private
+   */
+  getLocalizedStringAsync_(localizedStringId, elementToUse = this.element_) {
+    const languageCodes = this.getLanguageCodesForElement(elementToUse);
+
+    const localizedString = findLocalizedString(
+      this.localizedStringBundles_,
+      languageCodes,
+      localizedStringId
+    );
+    if (localizedString) {
+      return Promise.resolve(localizedString);
+    }
+    return new Promise((res) => {
+      const remove = this.bundleObserver_.add((languageCode) => {
```

Comment:
I don't understand why we need a Observable that fires for every added language. Couldn't we add a just a Promise map for pending languages? You could even take advantage of the `Signal` code to give this to you.

Reply:
It's true that in the meantime this can result in many observers firing multiple times (eventually we'll only install one language, the one in the doc), I'll try to update this code to be more efficient if I can think of a way.

There are some things to take into account:
1. Multiple bundles can be added to a language, eg: one for amp-story (async / fetched) and one for amp-story-auto-ads (sync for now) both in the same language, so a promise per language is not feasible.
2. When this migration is done, only one language will be installed so we can remove the notion of storing multiple languages on the localization service. **If I make this change first, it would make the observer path the best (I think) because you just need an observer or signal that there's new localization strings in the service.**

Maybe doing the change in 2 before merging this would make sense, so that having an observer for this is more suitable.

Reply:
Are the story and ads localizations strings totally separate? Could we make them different language keys in the map?

Reply:
Yes they are separate. The options are:

1) Your idea: It's possible to register them with different keys `amp-story` vs `amp-story-auto-ads` and then it would work, but definitely would require an intermediate PR to add this step in the code and update all usages of the localization service. **But** it's possible that at some point we bundle all of them in the same JSON so that the ads extension doesn't have to load 20 languages every time; and it would make this change useless.

2) New idea that I started to like a lot: We could change the signature of `registerLocalizedStringBundle` to support registering all the languages (bundled with amp-story) at the same time, and only fire the observer once at the end. We can do the same for the ads bundles, so the observer fires twice at all (but there will be about 50 things listening for it, which I don't know if it's bad or not). We can do `registerLocalizedStringBundles({'en': BundleEn, 'es': BundleEs})` (preferred) or `registerLocalizedStringBundle('en', BundleEn).registerLocalizedStringBundle('es', BundleEs).commit()` (less preferred).

3) Previous comment idea: We do what I said about first removing the notion of supporting multiple languages and just register the correct language (with the correct fallbacks).

Do you agree with idea 2, or wdyt?

Reply:
Does 2 solve the ads/story split? Eg, you're waiting on a string that comes from the ads bundle, you'd still need to wait for 2 promises?

Reply:
For 2 we don't need to keep track of what bundle is being installed, we'll just install the bundles as we're doing now, and bc the observer fires twice at all, each localization string that we want to translate will have to check if it's available at most 2 times (in reality won't happen most likely). So if the resource we're searching is not available when the observer fires the first time, we don't remove the handler yet and wait for the second one (the check is sth like `for each language in ['es-419', 'es', 'default'], chech if resource 53 is in the dict and return it, otherwise return null`).

I'll quickly note that in most cases the 20 languages that are currently installed get installed before the localization strings are required (this is what happens currently without the async behavior) so the observer mostly gets new handlers when the localizations come from the CDN JSONs (when they get installed late).

If this gets too complex for a Github thread I'd be happy to meet.

Reply:
I've written the changes required for this and they are 10 lines in index.js and a lot of tests that need to be updated:
https://github.com/ampproject/amphtml/pull/37870/files/53eb702dba28f90a87a613f040e5bcb7059e4408..d6cea31788d6a4ed1e1c8e25eba27784c910d7ca

It solves the issue of the observer firing many times (it only fires at most twice), no need to worry about the `auto-ads` split, and everything works as intended.